fpdcc/document-search
# Assorted bug fixes

## Overview

This PR fixes four small bugs:

- Make sure the `Survey.description` column can be correctly sorted (#46)
- Don't overwrite files when uploading files with identical names (https://github.com/fpdcc/document-search/issues/47)
- Allow partial name matches when searching (https://github.com/fpdcc/document-search/issues/48)
- Add `FlatDrawing.building_id` attribute (https://github.com/fpdcc/document-search/issues/49)

## Testing Instructions

### 1. Recreate your dev environment

* Update your development settings:

```
blackbox_cat configs/local_settings.dev.py.gpg > docsearch/local_settings.dev.py
```

* Take down your environment with `docker-compose down --volumes`
* Run `docker-compose up --build`
* In another shell, create a superuser: 

```
docker-compose run --rm app ./manage.py createsuperuser
```

* Load the app data:

```
docker-compose -f docker-compose.yml -f data/docker-compose.yml run --rm app make all
```

* Update the search index: 

```
docker-compose run --rm app ./manage.py rebuild_index
```

### 2. Test `Survey.description` sorting

- Navigate to http://localhost:8000/surveys/search/ and confirm that the `description` column sorts as you would expect
    - Note that since some descriptions are numbers, but the field overall is text, you may see some weird sorting behavior, e.g. `10` coming before `2` when sorting ascending

### 3. Test file overwriting

- Navigate to http://localhost:8000/surveys/1/ and download the survey PDF
- Create a new Survey and upload the PDF you just downloaded
- Confirm that the filename has a hash appended to the end and is not the same as the filename for the original PDF

### 4. Test fuzzy search

- Navigate to http://localhost:8000/flatdrawings/search/
- Confirm you can search for `cerm` or `cerm ak` and get results with the string `Cermak`
- Confirm you can search for `"cermak"` (with quotes) and get results with the exact match `Cermak`

### 5. Test `FlatDrawing.building_id` attribute

- Visit http://localhost:8000/flatdrawings/search/ and confirm:
    - There is a `Building ID` column in the table
    - There is a `Building ID` facet in the filters
    - You can sort by `Building ID`
- Navigate to http://localhost:8000/flatdrawings/8001/ and confirm the `Building ID` field is in the metadata
- Edit the flat drawing at http://localhost:8000/flatdrawings/update/8001/ and edit the `Building ID` field to a value like `1`
- Go back to search results and confirm that the new building ID is available in facets, and that when you order by building ID you see the row you edited first
